common/filesystem.h

```diff
@@ -0,0 +1,1374 @@
+auto MakeTmpDir() -> ErrorOr<RemovingDir, Error>;
+
+// Class modeling a file (or directory) status information structure.
+//
+// This provides a largely-portable model that callers can use, as well as a few
+// APIs to access non-portable implementation details when necessary.
+class FileStatus {
+ public:
+  // The size of the file in bytes.
+  auto size() const -> int64_t { return stat_buf_.st_size; }
+
+  auto type() const -> FileType {
+    return static_cast<FileType>(stat_buf_.st_mode &
+                                 static_cast<ModeType>(FileType::Mask));
+  }
+
+  // Convenience predicates to test for specific values of `type()`.
+  auto is_dir() const -> bool { return type() == FileType::Directory; }
+  auto is_file() const -> bool { return type() == FileType::RegularFile; }
+  auto is_symlink() const -> bool { return type() == FileType::SymbolicLink; }
+
+  // The read, write, and execute permissions for user, group, and others.
+  //
+  // These are represented using the Unix-style bit pattern that facilitates
+  // octal modeling:
+  // - Owner bit mask: 0700
+  // - Group bit mask: 0070
+  // - All bit mask:   0007
+  //
+  // For each, read is an octal value of `1`, write `2`, and execute `4`.
+  //
+  // Windows gracefully degrades to the effective permissions modeled using
+  // these values.
+  auto permissions() const -> ModeType { return stat_buf_.st_mode & 0777; }
+
+  // Non-portable APIs only available on Unix-like systems. See the
+  // documentation of the Unix `stat` structure fields they expose for their
+  // meaning.
+  auto inode() const -> uint64_t { return stat_buf_.st_ino; }
+  auto uid() const -> uid_t { return stat_buf_.st_uid; }
+
+ private:
+  friend DirRef;
+  friend Internal::FileRefBase;
+
+  FileStatus() = default;
+
+  struct stat stat_buf_ = {};
+};
+
+// The base class defining the core `File` API.
+//
+// While not used directly, this is the base class used to implement all of the
+// main `File` types: `ReadFileRef`, `WriteFileRef`, and `ReadWriteFileRef`.
+//
+// Objects using this type have access to an open file handle to a specific file
+// and expose operation on that open file. These operations may fail directly
+// with their `ErrorOr` return, but some errors may be deferred until the
+// underlying owning file is closed.
+//
+// The type provides reference semantics to the underlying file, but is
+// rebindable, movable, and copyable unlike a C++ language reference.
+class Internal::FileRefBase {
+ public:
+  // This object can be default constructed, but will hold an invalid file
+  // handle in that case. This is to support rebinding operations.
+  FileRefBase() = default;
+
+  // These objects are movable and copyable.
+  FileRefBase(FileRefBase&&) = default;
+  FileRefBase(const FileRefBase&) = default;
+  auto operator=(FileRefBase&&) -> FileRefBase& = default;
+  auto operator=(const FileRefBase&) -> FileRefBase& = default;
+
+  // Computers the file status.
```

Comment:
```suggestion
  // Computes the file status.
```

Though maybe, "Retrieves" or "Reads"?